### ISSUE TITLE
Make Theano support more different Production setup

### ISSUE THREAD

**Issue by nouiz** [Theano/Theano#2271]
This is a meta ticket with links to other tickets that would make Theano support more different production setup. Theano can already be used in different production setup.

- See this page with information: https://github.com/Theano/Theano/wiki/Servers
- [ ] Make Theano function threads safe gh-2186
- [ ] Generate a shared library per function that can be used from other non-python program: gh-2722.
  - [ ] Remove the NumPy and Python dependency from the library gh-2260
- [ ] Have a "production mode" for theano which removes most asserts and checks for errors gh-3441

How to for Theano in server mode:
- Using Theano in a server can be done.
    - In the VM/container, you still need g++ do be available. Make sure you compile the Theano function in the golden image. That way the c code will be reused in all clone of the VM/container.
    - To speed up the compilation, pickle and unpickle your Theano function in the same environment it will be used. With recent Theano version, this will skip the optimization phase. This don't allow to change cpu vs gpu computation and don't always support changing Theano version.

There is some examples somewhere of how to make call a Theano function from C via the Python C interface.

**Comment by Pranav0212** [comment]
Hey, I was just wondering if there was any update on this issue? Theano production would be really interesting especially for organizations looking to run Theano models on a server.

**Comment by nouiz** [comment]
not really. Do you have a particular setup in mind? Having Theano work on a
server is the easiest production setup and you can have it work as normal
Theano. You could pickle the Theano function and unpickle it to speed up
the compilation if you reuse always the same functions.

**Comment by Pranav0212** [comment]
Ah okay thanks :)

**Comment by ferrouswheel** [comment]
@nouiz Is there documentation somewhere that demonstrates this?

While I enjoy using Theano and related projects for research, the deployment story seemed awkward enough that I go with other frameworks.

Ideally the server deployed to shouldn't need a compiler and could be disabled from trying to compile anything, instead raising an Exception that the generated libraries are out of date.

(Sorry if this already exists or is explained somewhere else!)

**Comment by nouiz** [comment]
I updated the descriptions with more information and a link to other documentation. For pickle, it is simple:

```
import cPickle
f = theano.function(...)
with fd as open("fct.pkl", 'w'):
    cPickle.dump(f, fn, -1)
```
